### DISCUSSION TITLE
How to retrieve DataCollector tables from batch_run()?

### DISCUSSION THREAD

**Discussion by divilian** [mesa/mesa#1970] [Q&A]
The `DataCollector` constructor (Mesa 2.1.5) lets you specify three things: agent reporters, model reporters, and tables.

If I use the `batch_run()` function, it returns me a list, which when converted to a `DataFrame`, gives the output from all the agent reporters and all the model reporters.

But how do I get the output from the tables?

**Comment by rht** [comment]
> But how do I get the output from the tables?

`model.datacollector.get_table_dataframe("your_table_name")`

**Reply by divilian** [reply]
I don't have a `model` object, though. I only have a `Society` class (which extends `mesa.model`) which I pass to the batch_run() function:

```
batch_results = batch_run(Society,
    parameters={ 'args': args },
    iterations=args.num_sims,
    max_steps=args.max_iter,
    number_processes=None,
    data_collection_period=1
)

batch_results = pd.DataFrame(batch_results)
```

Tell me if I'm wrong, but at this point, `batch_run()` has done its work, having instantiated many models, and there isn't any way to get the tables back that I wrote to.

(Hope I'm wrong!)

**Reply by rht** [reply]
You are right that `batch_run()` only collects from agent and model reporter, but not from the tables: https://github.com/projectmesa/mesa/blob/f9798ebc754c7eaace4263046a9b97ab50073fed/mesa/batchrunner.py#L176-L191.

**Reply by tpike3** [reply]
Hmmmm..... that is probably a feature we need to add, if I remember correctly the problem was naming conventions. Regardless, off the top of my head would be in your model can you add a `df.to_csv("run_name.csv")` then in your output folder you will get a list of tables for each model run.

**Reply by quaquel** [reply]
I see this as another reason to spend some time carefully thinking through how data collection in general, is handled. I also hear from some users that this is a key bottleneck for them and that they often end up writing their own custom solutions.

**Comment by divilian** [comment]
> Regardless, off the top of my head would be in your model can you add a `df.to_csv("run_name.csv")` then in your output folder you will get a list of tables for each model run.

Sorry if I'm being dense. When you say "`df`" that implies I already have the DataFrame in question. But how do I get that, if all I've been doing is dutifully calling `datacollector.add_table_row()`? I can't figure out how to actually get the DataFrame with all of those rows at the right time. (I can of course call `.get_table_dataframe()` on the datacollector, but I need to do that only after I know the sim run has completed. It's too early to do it at the point when I'm writing any individual row.)

**Reply by Corvince** [reply]
Yes it is meant inside your model. I don't know if you have a stop condition, but say you run your models for 100 steps. Then inside your model.step you could do

```python
if model.schedule.steps == 100:
    df= model.datacollector.get_table()
    df.to_csv(file.csv)
```

**Comment by divilian** [comment]
Yeah, unfortunately the stop condition isn't that simple, and can't be known ahead of time. What I need to do is stuff information in a table every time a particular kind of event occurs, and then retrieve that table at simulation's end.

This all works great for model and agent reporters! Just not tables. :(

**Reply by quaquel** [reply]
How do you know a sim run is completed? There must be some logic in your model or it is based on ticks. In either case, you should be able to store the results as suggested.

To be clear: this should be fixed in MESA itself. Just trying to find a solution for you in the meantime.

**Comment by divilian** [comment]
> How do you know a sim run is completed? There must be some logic in your model or it is based on ticks. In either case, you should be able to store the results as suggested.

Ah, I see. Okay, that might work as a workaround.

Question, though: in my model class, at the point where it figures out that the sim is in fact completed, and time to save the dataframe as a csv file, how does it know the simulation "number" (or some other unique ID) of that particular simulation run in the suite? (Otherwise, all the simulations in the batch run would overwrite each others' results in their identically-named .csv file.)

@tpike3 mentioned saving the results as "`run_name.csv`" -- another way of phrasing my question is "what do I use for the 'run name' that @tpike3 refers to here?" (I assume that wasn't literal.)
